File: iPSC/examples/nx_c/pi/node.c
/***********************************************************************
*
*			Node Program
*
*	Approximate pi with the n-point rectangle quadrature rule
*	applied to the integral from 0 to 1 of 4 / (1+x**2)
*
*	pi = h * sum from i = 1 to n of f(xi)
*
*	where  h = 1/n
*	       xi=(i-1/2)*h
*	       f(x)=4 / (1+x**2)
*
***********************************************************************/

#define intsize 2

int pid,cid,id,dim,type,cnt,root,rnid,rpid;
int copen(),cubedim(),mynode();
int n,p,i;
double x,sum,pi,h,work;

main() { /* node program */

	double f();

/*
	Open communications area
*
	Find node number, cube dimension, number of processors
*/

	pid = 0;
	cid = copen(pid);
        id = mynode();
	dim = cubedim();
	p = 1<<dim;

/*
	Use global receive to get n from the host.
*/

	n = 1;
	while(n > 0) {
	   type = 1;
	   recvw(cid,type,&n,intsize,&cnt,&rnid,&rpid);

/*
	   Compute this node's partial sum
	   Partial sums are interleaved
*/

	   if (n > 0) {
	      h = 1.0 / n;
	      sum = 0.0;
	      for (i=id+1; i<=n; i=i+p) {
	         x = (i-0.5) * h;
		 sum = sum + f(x);
	      }

	      pi = h * sum;

/*
	      Use global sum to assemble partial sums and send result to host
*/

	      type = 2;
	      root = 0x8000;
	      gop(cid,type,&pi,1,'+',root,dim,&work);
	   }
	}

}

double f(x)
double x;

{

  return(4/(1+x*x));

}

File: iPSC/examples/nx_rmf/pi/node.f
	PROGRAM NODE

C***********************************************************************
C
C			Node Program
C
C	Approximate pi with the n-point rectangle quadrature rule
C	applied to the integral from 0 to 1 of 4 / (1+x**2)
C
C	pi = h * sum from i = 1 to n of f(xi)
C
C	where  h = 1 / n
C	       xi = (i - 1 / 2)*
C	       f(x) = 4/ (1+ x**2)
C
C***********************************************************************

	INTEGER PID,CID,ID,DIM,TYPE,CNT,ROOT,RNID,RPID
	INTEGER COPEN, CUBEDIM, MYNODE
	INTEGER N,P,I,INTSIZE
	DOUBLE PRECISION F,X,SUM,PI,H,WORK

	PARAMETER (INTSIZE=4)

	F(X) = 4.D0/(1.D0 + X*X)

C
C	Open communications area
C
C	Find node number, cube dimension, number of processors
C

	PID = 0
	CID = COPEN(PID)
	ID = MYNODE()
	DIM = CUBEDIM()
	P = 2**DIM

C
C	Use global receive to get n from the host
C

10	TYPE = 1
	CALL RECVW (CID,TYPE,N,INTSIZE,CNT,RNID,RPID)

C
C	Compute this node's partial sum
C	Partial sums are interleaved
C

	IF (N .GT. 0) THEN
	   H = 1.0D0 / N
	   SUM = 0.0D0
	   DO 20 I = ID + 1, N , P
	      X = (I - 0.5D0)*H
	      SUM = SUM + F(X)
20	   CONTINUE

	   PI = H * SUM

C
C	   Use global sum to assemble partial sums and send result to host
C

	   TYPE = 2
	   ROOT = -32768
	   CALL GOP (CID,TYPE,PI,1,'+',ROOT,DIM,WORK)

	   GO TO 10

	END IF

	END
